plugins/workspace-minimap/src/focus_mode.ts

```diff
@@ -0,0 +1,171 @@
+      this.primaryWorkspace = primary;
+      this.minimapWorkspace = minimap;
+      this.id = this.minimapWorkspace.id;
+    }
+
+
+    /**
+     * Initializes focus mode.
+     */
+    init() {
+      // Make the svg group element.
+      this.svgGroup = Blockly.utils.dom.createSvgElement(
+          Blockly.utils.Svg.G, {'class': 'focusMode'}, null);
+
+      // Make the mask under the svg group.
+      const mask = Blockly.utils.dom.createSvgElement(
+          new Blockly.utils.Svg('mask'),
+          {'id': 'focusModeMask' + this.id},
+          this.svgGroup);
+
+      // Make the backround under the svg group.
+      this.background = Blockly.utils.dom.createSvgElement(
+          Blockly.utils.Svg.RECT, {
+            'x': 0,
+            'y': 0,
+            'width': '100%',
+            'height': '100%',
+            'mask': 'url(#focusModeMask' + this.id + ')',
+          }, this.svgGroup);
+
+      // Make the white layer under the svg mask.
+      Blockly.utils.dom.createSvgElement(
+          Blockly.utils.Svg.RECT, {
+            'x': 0,
+            'y': 0,
+            'width': '100%',
+            'height': '100%',
+            'fill': 'white',
+          }, mask);
+
+      // Make the black layer under the mask.
+      this.rect = Blockly.utils.dom.createSvgElement(
+          Blockly.utils.Svg.RECT, {
+            'x': 0,
+            'y': 0,
+            'rx': Blockly.Bubble.BORDER_WIDTH,
+            'ry': Blockly.Bubble.BORDER_WIDTH,
+            'fill': 'black',
+          }, mask);
+
+      // Theme.
+      this.background.setAttribute('fill', '#e6e6e6');
+
+      // Add the svg group to the minimap.
+      const parentSvg = this.minimapWorkspace.getParentSvg();
+      if (parentSvg.firstChild) {
+        parentSvg.insertBefore(this.svgGroup, parentSvg.firstChild);
+      } else {
+        parentSvg.appendChild(this.svgGroup);
+      }
+
+      window.addEventListener('resize', () => void this.update());
+      this.onChangeWrapper = this.onChange.bind(this);
+      this.primaryWorkspace.addChangeListener(this.onChangeWrapper);
+
+      this.update();
+    }
+
+
+    /**
+     * Disposes of the focus mode.
+     * Unlinks from all DOM elements and remove all event listeners
+     * to prevent memory leaks.
+     */
+    dispose() {
+      if (this.svgGroup) {
+        Blockly.utils.dom.removeNode(this.svgGroup);
+      }
+      if (this.onChangeWrapper) {
+        this.primaryWorkspace.removeChangeListener(this.onChangeWrapper);
+      }
+    }
+
+
+    /**
+     * Handles events triggered on the primary workspace.
+     * @param event The event.
+     */
+    private onChange(event: Blockly.Events.Abstract): void {
+      if (BlockEvents.has(event.type)) {
+        this.update();
+      }
+    }
+
+
+    /**
+     * Positions and sizes the highlight on the minimap
+     * based on the primary workspace.
+     */
+    private update(): void {
```

Comment:
Can you use a [destructuring assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) in this method to aliase the `width`s `height`s etc? It may make it more messy :P but I'm hoping it makes this a bit easier to read!

Give it a shot and use your judgement :P

(I'm hoping it lets you conform to the [rectangle rule](https://github.com/google/google-java-format/wiki/The-Rectangle-Rule)).

Reply:
So I created aliases for each metric category but I couldn't find a way to use destructuring to really improve readability.

Reply:
Looks great!